pkg/controller/networkpolicyrecommendation/controller.go

```diff
@@ -168,11 +300,420 @@ func (c *NPRecommendationController) syncNPRecommendation(key apimachinerytypes.
+	return c.updateNPRecommendationStatus(
+		npReco,
+		crdv1alpha1.NetworkPolicyRecommendationStatus{
+			State:           crdv1alpha1.NPRecommendationStateRunning,
+			CompletedStages: completedStages,
+			TotalStages:     totalStages,
+		},
+	)
+}
+
+func (c *NPRecommendationController) checkSparkApplicationStatus(npReco *crdv1alpha1.NetworkPolicyRecommendation) (string, error) {
+	if npReco.Status.SparkApplication == "" {
+		return "", c.updateNPRecommendationStatus(
+			npReco,
+			crdv1alpha1.NetworkPolicyRecommendationStatus{
+				State:    crdv1alpha1.NPRecommendationStateFailed,
+				ErrorMsg: "Spark Application should be started before status checking",
+			},
+		)
+	}
+
+	state, errorMessage, err := getPolicyRecommendationStatus(c.kubeClient, npReco.Status.SparkApplication, npReco.Namespace)
+	if err != nil {
+		return state, err
+	}
+	klog.V(4).InfoS("Got Spark Application state", "state", state, "NetworkRecommendationPolicy", npReco.Name)
+	if state == "RUNNING" {
+		return state, c.updateNPRecommendationStatus(
+			npReco,
+			crdv1alpha1.NetworkPolicyRecommendationStatus{
+				State:    crdv1alpha1.NPRecommendationStateRunning,
+				ErrorMsg: errorMessage,
+			},
+		)
+	} else if state == "COMPLETED" {
+		return state, c.updateNPRecommendationStatus(
+			npReco,
+			crdv1alpha1.NetworkPolicyRecommendationStatus{
+				State:    crdv1alpha1.NPRecommendationStateCompleted,
+				ErrorMsg: errorMessage,
+			},
+		)
+	} else if state == "FAILED" || state == "SUBMISSION_FAILED" || state == "FAILING" || state == "INVALIDATING" {
+		return state, c.updateNPRecommendationStatus(
+			npReco,
+			crdv1alpha1.NetworkPolicyRecommendationStatus{
+				State:    crdv1alpha1.NPRecommendationStateFailed,
+				ErrorMsg: fmt.Sprintf("policy recommendation job failed, state: %s, error message: %v", state, errorMessage),
+			},
+		)
+	}
+	return state, nil
+}
+
+func (c *NPRecommendationController) startJob(npReco *crdv1alpha1.NetworkPolicyRecommendation) error {
+	// Validate Cluster readiness
+	if err := validateCluster(c.kubeClient, npReco.Namespace); err != nil {
+		return err
+	}
+	err := c.startSparkApplication(npReco)
+	// Mark the NetworkPolicyRecommendation as failed and not retry if it failed due to invalid request
+	if err != nil && strings.Contains(err.Error(), "invalid request") {
+		return c.updateNPRecommendationStatus(
+			npReco,
+			crdv1alpha1.NetworkPolicyRecommendationStatus{
+				State:    crdv1alpha1.NPRecommendationStateFailed,
+				ErrorMsg: fmt.Sprintf("error in creating NetworkPolicyRecommendation: %v", err),
+			},
+		)
+	}
+	// Schedule periodical resync for successful starting
+	if err == nil {
+		c.addPeriodicSync(apimachinerytypes.NamespacedName{
+			Name:      npReco.Name,
+			Namespace: npReco.Namespace,
+		})
+	}
+	return err
+}
+
+func (c *NPRecommendationController) startSparkApplication(npReco *crdv1alpha1.NetworkPolicyRecommendation) error {
```

Comment:
can we abstract these validation logic to some querrier-like interface so it can be shared with API server REST handler? This way we can move the error reporting at REST CREATE time. Otherwise a success response from API server could end up in invalid state due to user input error, which looks confusing.  cc @yuntanghsu

Reply:
As Yanjun pointed out offline, as the similar checks are also done on CLI, and no other component / user is expected to be directly calling the API server at the moment, validation at REST handler doesn't provide much benefit at the moment. probably can skip this at the moment until when UI is to be added and we can reevaluate.

Reply:
Will the user directly create a NPR CR to start the SparkApplication? If no, maybe we can only validate the input on CLI side.

Reply:
We do not suggest user to do this. But we do not prevent user from doing this. I think the validation here could be a protection in case this happens.